components/aws/sagemaker/rlestimator/src/sagemaker_rlestimator_component.py

```diff
@@ -0,0 +1,275 @@
+            INFO:root: - Overtraining: NoIssuesFound
+            ERROR:root:- CustomGradientRule: Error
+            If last_print is True:
+            INFO:root: - LossNotDecreasing: IssuesFound
+            INFO:root:   - RuleEvaluationConditionMet: Evaluation of the rule LossNotDecreasing at step 10 resulted in the condition being met
+
+        Args:
+            response: A describe training job response.
+            last_print: If true, prints each of the debug rule issues if found.
+        """
+        for debug_rule in response["DebugRuleEvaluationStatuses"]:
+            line_ending = "\n" if last_print else ""
+            if "StatusDetails" in debug_rule:
+                status_details = (
+                    f"- {debug_rule['StatusDetails'].rstrip()}{line_ending}"
+                )
+                line_ending = ""
+            else:
+                status_details = ""
+            rule_status = f"- {debug_rule['RuleConfigurationName']}: {debug_rule['RuleEvaluationStatus']}{line_ending}"
+            if debug_rule["RuleEvaluationStatus"] == "Error":
+                log_fn = logging.error
+                status_padding = 1
+            else:
+                log_fn = logging.info
+                status_padding = 2
+
+            log_fn(f"{status_padding * ' '}{rule_status}")
+            if last_print and status_details:
+                log_fn(f"{(status_padding + 2) * ' '}{status_details}")
+        self._print_log_header(50)
+
+    def _after_job_complete(
+        self,
+        job: object,
+        request: Dict,
+        inputs: SageMakerRLEstimatorInputs,
+        outputs: SageMakerRLEstimatorOutputs,
+    ):
+        outputs.job_name = self._rlestimator_job_name
+        outputs.model_artifact_url = self._get_model_artifacts_from_job(
+            self._rlestimator_job_name
+        )
+        outputs.training_image = self._get_image_from_job(self._rlestimator_job_name)
+
+    def _on_job_terminated(self):
+        self._sm_client.stop_training_job(TrainingJobName=self._rlestimator_job_name)
+
+    def _print_logs_for_job(self):
+        self._print_cloudwatch_logs(
+            "/aws/sagemaker/TrainingJobs", self._rlestimator_job_name
+        )
+
+    def _create_job_request(
+        self, inputs: SageMakerRLEstimatorInputs, outputs: SageMakerRLEstimatorOutputs,
+    ) -> RLEstimator:
+        # Documentation: https://sagemaker.readthedocs.io/en/stable/frameworks/rl/sagemaker.rl.html
+        # We need to configure region and it is not something we can do via the RLEstimator class.
+
+        # Only use max wait time default value if electing to use spot instances
+        if not inputs.spot_instance:
+            max_wait_time = None
+        else:
+            max_wait_time = inputs.max_wait_time
+
+        estimator = RLEstimator(
+            entry_point=inputs.entry_point,
+            source_dir=inputs.source_dir,
+            image_uri=SpecInputParsers.nullable_argument(inputs.image),
```

Comment:
These input parsers are intended to be used as the `input_type` for the input validators. What is the advantage of using it like this?

Reply:
When instantiating the RLEstimator class these values are validated, and need to be of an expected type, or None. The component compiler will convert some empty values to an empty string: https://github.com/kubeflow/pipelines/blob/ccd0f2571be2b212f6ff742bf5d9f729bc5f0aad/components/aws/sagemaker/common/component_compiler.py#L151
This is a way of allowing the component to provide empty values and then catching them and returning None at instantiation time.

Reply:
Hmm I don't remember why this `elif` statement exist. I think to maintain backward compatibility with the older logic in which we use empty strings to denote empty input variables. That's a shame, they really should be `None`.

Reply:
Updated as per recommendation